src/selectors/publish.js

```diff
@@ -146,28 +155,47 @@ export const getSanitizedProfileData: Selector<
   }
 );
 
+/**
+ * The blob is needed for both the download size, and the ObjectURL.
+ */
 export const getSanitizedProfileBlob: Selector<Promise<Blob>> = createSelector(
   getSanitizedProfileData,
   async profileData =>
     new Blob([await profileData], { type: 'application/octet-binary' })
 );
 
-export const getCompressedProfileBlobUrl: Selector<
+// URL.createObjectURL are not GCed, they must be cleaned up manually. Store a reference
+// to the previous object URL here. Only retain the latest one.
+let _previousObjectUrl;
+
+/**
+ * This selector generates the string that can be downloaded, or uploaded for persisting
+ * profiles.
+ */
+export const getCompressedProfileObjectUrl: Selector<
   Promise<string>
-> = createSelector(getSanitizedProfileBlob, blobPromise =>
-  blobPromise.then(blob => URL.createObjectURL(blob))
-);
+> = createSelector(getSanitizedProfileBlob, async blobPromise => {
+  const blob = await blobPromise;
+  if (_previousObjectUrl) {
+    // Make sure and clean up the previously created object URL so that we
+    // don't leak it.
+    // https://developer.mozilla.org/en-US/docs/Web/API/URL/revokeObjectURL
+    URL.revokeObjectURL(_previousObjectUrl);
+  }
+  _previousObjectUrl = URL.createObjectURL(blob);
+  return _previousObjectUrl;
+});
```

Comment:
I really prefer that we create and revoke in the location where we use it. Here this is very disconnected from its usage.

There's a risk of a race condition too: when revoking the previous URL, in theory you don't know where it's used, and so there's a risk that it's used after the revocation.

Admittedly the risk is super small in the current code and that's why I won't block the PR for this.

Also I realize that with the approach you took the memory won't be recovered once the user merely opened the publish panel. So we'll have both the current profile and the sanitized profile in memory, forever, until the user closes the page. This can be handled in a follow-up.

Reply:
I'll follow-up with a component that handles this.